### PR TITLE
Add experimental support of fused decoder layer for llama2

### PR DESCRIPTION
**Pull request by plusbang**
## Description

Refactor https://github.com/intel-analytics/ipex-llm/pull/11716

This experimental support is only verified on llama2-7b currently.

- To avoid influence on GPU support and NPU support for other models (llama3, qwen2, etc), add `DynamicFusedNormalCache`, `llama_fused_model_forward` and `pipeline_parallel` in `transformers/npu_models/`. Then only specify `pipeline_parallel_stages > 1` will use related logic.

- Add experimental example and readme

### 4. How to test?
- [x] Unit test: Please manually trigger the PR Validation [here](https://github.com/intel-analytics/ipex-llm-workflow/actions/workflows/llm-PR-validation.yml) by inputting the PR number (e.g., `1234`). And paste your action link here once it has been successfully finished.
https://github.com/intel-analytics/ipex-llm-workflow/actions/runs/10362835964
- [x] Application test

### CODE REVIEW

File: python/llm/src/ipex_llm/transformers/npu_model.py [hunk 1]
```diff
@@ -131,13 +131,49 @@ def from_pretrained(cls,
...
+                cls.load_convert(qtype, model, 'cpu', *args, **kwargs)
+                create_npu_kernels(model)
+                model = model.eval()
+            else:
+                cls.load_convert(qtype, model.model, 'cpu', *args, **kwargs)
+                create_npu_kernels(model.model)
+                model = model.eval()
+                model.model.embed_tokens.to(torch.float32)
+                model.model.norm.to(torch.float32)
+                model.lm_head.to(torch.float32)
```
[the remark this answers is not in the document]
Comment (yangw1234) [reply python/llm/src/ipex_llm/transformers/npu_model.py:159]: maybe we should wrap this logic to a function similar to `optimize_llm`

File: python/llm/src/ipex_llm/transformers/npu_models/pipeline_parallel.py
Comment (yangw1234) [inline python/llm/src/ipex_llm/transformers/npu_models/pipeline_parallel.py:1]: I think the changes in `pipeline_parallel.py`, if cleaned up a bit , should not affect gpu code. Maybe we should try to merge them later.

**Review by yangw1234 — APPROVED** [review]

File: python/llm/example/NPU/HF-Transformers-AutoModels/LLM/README.md [hunk 2]
```diff
@@ -81,3 +81,62 @@ Inference time: xxxx s
...
+We suggest using conda to manage environment:
+```bash
+conda create -n llm python=3.10
+conda activate llm
+
+# install ipex-llm with 'all' option
+pip install --pre --upgrade ipex-llm[all]
+pip install --pre --upgrade bigdl-core-npu
+
+pip install transformers==4.40
```
Comment (yangw1234) [inline python/llm/example/NPU/HF-Transformers-AutoModels/LLM/README.md:98]: I am experimenting on `transformers==3.39.3`, have we verified 4.40?

Comment (plusbang) [reply python/llm/example/NPU/HF-Transformers-AutoModels/LLM/README.md:98]: > I am experimenting on `transformers==3.39.3`, have we verified 4.40?

yes, I have also verified `transformers==4.40`. 

Have added `optimize_llm_post` to wrap `lm_head` related logic etc. And considering we may need additional process to address prefill, maybe we could keep separating npu and gpu PP temporarilly.

**Comment by plusbang** [comment]
Merge it first, please let me know if any suggestion or comment.

**Pull request merged**